src/rgw/rgw_pubsub.cc

```diff
@@ -462,13 +482,74 @@ std::string rgw_pubsub_dest::to_json_str() const
   return ss.str();
 }
 
+void rgw_pubsub_dest::decode_json(JSONObj* f) {
+  using rgw::notify::DEFAULT_CONFIG;
+  using rgw::notify::DEFAULT_GLOBAL_VALUE;
+  JSONDecoder::decode_json("push_endpoint", push_endpoint, f);
+  JSONDecoder::decode_json("push_endpoint_args", push_endpoint_args, f);
+  JSONDecoder::decode_json("push_endpoint_topic", arn_topic, f);
+  JSONDecoder::decode_json("stored_secret", stored_secret, f);
+  JSONDecoder::decode_json("persistent", persistent, f);
+  std::string ttl;
+  JSONDecoder::decode_json("time_to_live", ttl, f);
+  time_to_live = ttl == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE : std::stoul(ttl);
+
+  std::string max_retry;
+  JSONDecoder::decode_json("max_retries", max_retry, f);
+  max_retries = max_retry == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE
+                                            : std::stoul(max_retry);
+
+  std::string sleep_dur;
+  JSONDecoder::decode_json("retry_sleep_duration", sleep_dur, f);
+  retry_sleep_duration = sleep_dur == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE
+                                                     : std::stoul(sleep_dur);
+}
+
 RGWPubSub::RGWPubSub(rgw::sal::Driver* _driver, const std::string& _tenant)
   : driver(_driver), tenant(_tenant)
 {}
 
 int RGWPubSub::read_topics(const DoutPrefixProvider *dpp, rgw_pubsub_topics& result, 
     RGWObjVersionTracker *objv_tracker, optional_yield y) const
 {
+  if (driver->get_zone()->get_zonegroup().supports_feature(
+          rgw::zone_features::notification_v2)) {
+    void* handle = NULL;
+    auto ret =
+        driver->meta_list_keys_init(dpp, "topic", std::string(), &handle);
+    if (ret < 0) {
+      return ret;
+    }
+    bool truncated;
+    int max = 1000;
+    do {
+      std::list<std::string> topics;
+      ret = driver->meta_list_keys_next(dpp, handle, max, topics, &truncated);
+      if (ret < 0) {
+            ldpp_dout(dpp, 1)
+                << "ERROR: lists_keys_next(): " << cpp_strerror(-ret) << dendl;
+            break;
+      }
+      for (auto& topic_entry : topics) {
+        std::string topic_name;
+        std::string topic_tenant;
+        parse_topic_entry(topic_entry, &topic_tenant, &topic_name);
+        if (tenant != topic_tenant) {
+          continue;
```

Comment:
if we're only trying to list topics of a given `tenant` name, we can do a prefixed search instead of filtering over all keys. something like this?
```diff
     auto ret =
-        driver->meta_list_keys_init(dpp, "topic", std::string(), &handle);
+        driver->meta_list_keys_init(dpp, "topic", tenant + ":", &handle);
```
if we change `get_topic_key()` to return `"{tenant}:{name}"` even when tenant is empty, that would let us list only the global tenant's topics with `prefix = ":"`

Reply:
oops sorry, that `std::string()` argument is the marker, not the prefix. we'd need to use something rados-specific to do the prefix filtering. that could be exposed in a separate `sal::Driver::list_topics_v2()` interface

this function serves the https://docs.aws.amazon.com/sns/latest/api/API_ListTopics.html API, so i think it's important that we filter efficiently and paginate the results so we don't have to read all topics into memory. this v2 implementation has to read each entry from rados (or cache), so it's important that we limit how much we'll do in a single http request

i've started rebasing https://github.com/ceph/ceph/pull/54777 onto this. i'm happy to make these changes in a separate pr

Reply:
according to that [ListTopics](https://docs.aws.amazon.com/sns/latest/api/API_CreateTopic.html) api, all we have to return is "A list of topic ARNs", and we could reconstruct those ARNs from the metadata keys alone

however, i see that our ListTopicsResult response includes a lot more than just TopicArn:
https://github.com/ceph/ceph/blob/fb466aa51d42382f1642aec99fc85482275527b4/src/rgw/rgw_pubsub.cc#L343-L351

@yuvalif would we break anything if RGWPSListTopicsOp only returned the TopicArns?

Reply:
@cbodley  `encode_xml` is also used for Sns:GetTopicAttributes, so i am guessing that could be the reason even the list_topics is sending all the data.

Reply:
in the current implementation "ListTopics" is the same as "GetTopicAttribute" (only for all topics).
we haven't had any complaints that this is not according to spec. 
so, unless there is a reason, its probably better to not change for now.